walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,76 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;	
 import java.util.Collection;
+import java.util.Collections;	
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if(request.getOptionalAttendees().size() == 0) {
+        return getOpenSlots(events, request);	
+    }	
+
+    Collection<String> allAttendees = new ArrayList<String>();	
+    allAttendees.addAll(request.getAttendees());	
+    allAttendees.addAll(request.getOptionalAttendees());	
+    MeetingRequest allAttendeesRequest = new MeetingRequest(allAttendees, request.getDuration());	
+    Collection<TimeRange> openSlotsForMandatoryAttendees = getOpenSlots(events, request);	
+    Collection<TimeRange> openSlotsForAllAttendees = getOpenSlots(events, new MeetingRequest(allAttendees, request.getDuration()));	
+
+    if(openSlotsForAllAttendees.size() == 0 && request.getAttendees().size() > 0) {	
+      return openSlotsForMandatoryAttendees;	
+    } else {	
+      return openSlotsForAllAttendees;	
+    }	
+  }	
```

Comment:
If openSlotsForAllAttendees is non empty, we don't need to compute openSlotsForMandatoryAttendees. This part can be refactored to avoid not necessary computation.